Add console plugin:install command

jsantos42 opened the pull request:
### Description:
Fixes #21003 

(Ref DEV-16969)

### Review

* [x] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [x] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [x] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [x] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [x] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [x] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [x] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [x] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [x] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [x] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [x] Existing documentation updated if needed

jsantos42 committed:
Add console plugin:install command

sgiehl commented:
Hey @jsantos42.
Thanks for providing this Pull Request. I had a rough look through the code and by the looks the new command actually doesn't fully do what would be expected.
It only seems to create the `plugin.json` file(s), but doesn't fetch the actual plugin content.

jsantos42 committed:
fix: install the remaining files of the plugin

jsantos42 commented:
@sgiehl you were right, just refactored the code in order to add the remaining files of the plugin. Also added some PHPDoc notations.
I do have one question, would you still keep the specific exception on line 53, given that I have to catch the no connection exception with a general exception?

michalkleiner commented:
@jsantos42 thanks for putting the effort into this, it's very appreciated. Can I ask why do you think we need to write the plugin.json file first? Wouldn't the installation take care of that anyway?

jsantos42 committed:
fix: remove redundant code

sgiehl commented on plugins/CorePluginsAdmin/Commands/InstallPlugin.php:
```diff
@@ -0,0 +1,69 @@
+                $this->installPlugin($pluginName);
+                $output->writeln(sprintf("Installed plugin <info>%s</info>", $pluginName));
+            } catch (\Piwik\Plugins\CorePluginsAdmin\PluginInstallerException $e) {
+                $output->writeln(sprintf("<error>%s</error>", $e->getMessage()));
```
In this case it might be good to also show a message like `Unable to install plugin XXXX: `.
Otherwise if you e.g. try to install multiple plugins and more than one plugin has an error it is not directly visible, which error message was for which plugin. Even though it's always in the provided order.

![image](https://github.com/user-attachments/assets/45e45248-afa1-44f8-aa0c-8bf7d38fe329)

sgiehl commented on plugins/CorePluginsAdmin/Commands/InstallPlugin.php:
```diff
@@ -0,0 +1,69 @@
+    /**
+     * @param string $pluginName
+     */
+    private function installPlugin($pluginName): void
```
```suggestion
    private function installPlugin(string $pluginName): void
```

sgiehl commented on plugins/CorePluginsAdmin/Commands/InstallPlugin.php:
```diff
@@ -0,0 +1,69 @@
+                continue;
+            }
+
+            if (!$pluginManager->isPluginInFilesystem($pluginName) && !Marketplace::isMarketplaceEnabled()) {
```
What is the `isPluginInFilesystem` check here meant for?
If Marketplace is disabled I guess the commend shouldn't be doing anything at all, right? So guess the `isMarketplaceEnabled` check could be done before iterating over `$pluginName`

sgiehl reviewed (changes requested):
Left some minor comments. Otherwise the command generally works as expected now.

jsantos42 committed:
fix: mention implicated plugin in error message

jsantos42 committed:
refactor: add function parameter type

jsantos42 committed:
refactor: move isMarketplaceEnabled check out of the loop

jsantos42 committed:
fix: fix plugin not being listed in the plugin manager

michalkleiner commented on plugins/CorePluginsAdmin/Commands/InstallPlugin.php:
```diff
@@ -0,0 +1,73 @@
+            }
+
+            try {
+                $this->installPlugin($pluginName, $pluginManager);
```
You don't need to pass in the plugin manager instance, you can call `Manager::getInstance()` in the `installPlugin` method, it uses the static container and it will give you the existing instance. It's easier and simpler to only pass in the plugin name.

jsantos42 replied:
That's definitely cleaner. What if we promote it to a class property? It's being used in 2 methods...if you don't like this idea, I can always submit it as you said

michalkleiner replied:
That's why we have the static container, we don't need to store it in a property, it does that for us.

jsantos42 replied:
Got it. Just fixed that and rebased the code

jsantos42 committed:
refactor: promote pluginManager to class property

jsantos42 committed:
Revert "refactor: promote pluginManager to class property"

This reverts commit e38d349a955e40fb32c6bdd4bc4f45941b685b2a.

jsantos42 committed:
refactor: simplify method signature

jsantos42 committed:
fix: improve error message when plugin is not downloadable

jsantos42 committed:
chore: fix code indentation

jsantos42 committed:
refactor: use dependency injection for PluginInstaller

jsantos42 committed:
feat: make sure config file is writable

jsantos42 committed:
chore: update CHANGELOG file

jsantos42 committed:
feat: make plugin:install command able to also update the plugin

tsteur commented:
Awesome, thanks @jsantos42

sgiehl commented on plugins/CorePluginsAdmin/Commands/InstallPlugin.php:
```diff
@@ -0,0 +1,71 @@
+        foreach ($pluginNames as $pluginName) {
+            try {
+                $this->installPlugin($pluginName);
+                $output->writeln(sprintf("Installed plugin <info>%s</info>", $pluginName));
```
@jsantos42 We could make a difference here between installing and updating. If the plugin was installed before, the plugin might have been updated. And if we want to be super accurate we could check if the installed version actually changed. If it doesn't actually nothing happened as there was no update available.
If you don't have time to put some extra effort into this, it's also fine to just update the text to `Installed or updated plugin...` for now.
Otherwise this PR looks good and is almost ready to merge

sgiehl replied:
```suggestion
                $output->writeln(sprintf("Installed or updated plugin <info>%s</info>", $pluginName));
```

sgiehl replied:
I've now changed the text only, so we can get that merged. Feel free to create a follow up PR if you want.

jsantos42 replied:
@sgiehl sorry for getting back to you only now, I was on vacation. Thank you for fixing that on my behalf! And yes, I might address that on a future PR indeed  🙌

sgiehl commented on plugins/CorePluginsAdmin/Commands/InstallPlugin.php:
```diff
@@ -0,0 +1,71 @@
+                $this->installPlugin($pluginName);
+                $output->writeln(sprintf("Installed plugin <info>%s</info>", $pluginName));
+            } catch (\Piwik\Plugins\CorePluginsAdmin\PluginInstallerException $e) {
+                $output->writeln(sprintf("<error>Unable to install plugin %s. %s</error>", $pluginName, $e->getMessage()));
```
```suggestion
                $output->writeln(sprintf("<error>Unable to install or update plugin %s. %s</error>", $pluginName, $e->getMessage()));
```

sgiehl committed:
Apply suggestions from code review
